spring-data-relational/src/main/java/org/springframework/data/relational/core/mapping/BasicRelationalPersistentProperty.java

```diff
@@ -173,14 +188,22 @@ public int getSqlType() {
 
 	@Override
 	public String getReverseColumnName() {
-		return columnName.get().orElseGet(() -> context.getNamingStrategy().getReverseColumnName(this));
+		return collectionIdColumnName.get().orElseGet(
+				() -> columnName.get().orElseGet(
+						() -> context.getNamingStrategy().getReverseColumnName(this)
+				)
+		);
 	}
 
 	@Override
 	public String getKeyColumn() {
 
 		if (isQualified()) {
-			return keyColumnName.get().orElseGet(() -> context.getNamingStrategy().getKeyColumn(this));
+			return collectionKeyColumnName.get().orElseGet(
```

Comment:
This resolution (relation to `keyColumnName`) should rather go into `collectionKeyColumnName ` resolution to keep the fallback code more closely together.